Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+	{
+		if ((m_pMyUpgradeInv[i] != nullptr) && (m_pMyUpgradeInv[i]->pUIIcon == pUIIcon))
+			return m_pMyUpgradeInv[i];
+	}
+	if (m_pUpgradeItemSlot && m_pUpgradeItemSlot->pUIIcon == pUIIcon)
+		return m_pUpgradeItemSlot;
+
+	return nullptr;
+}
+
+void CUIItemUpgrade::Open()
+{
+	SetVisible(true);
+	ItemMoveFromInvToThis();
+
+	GoldUpdate();
+}
+
+void CUIItemUpgrade::GoldUpdate()
+{
+	if (m_pStrMyGold)
+	{
+		m_pStrMyGold->SetStringAsInt(CGameBase::s_pPlayer->m_InfoExt.iGold);
+	}
+}
+
+void CUIItemUpgrade::ItemMoveFromInvToThis()
+{
+	CUIInventory* pInven = CGameProcedure::s_pProcMain->m_pUIInventory;
+	if (!pInven)
+		return;
+
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (pInven->m_pMyInvWnd[i])
+		{
+			__IconItemSkill* spItem = pInven->m_pMyInvWnd[i];
+			spItem->pUIIcon->SetParent(this);
+
+			pInven->m_pMyInvWnd[i] = nullptr;
+			CN3UIArea* pArea;
+
+			char szID[32];
+			sprintf(szID, "a_slot_%d", i);
+			pArea = (CN3UIArea*)GetChildByID(szID);
+			if (pArea)
+			{
+				spItem->pUIIcon->SetRegion(pArea->GetRegion());
+				spItem->pUIIcon->SetMoveRect(pArea->GetRegion());
+			}
+
+			m_pMyUpgradeInv[i] = spItem;
+		}
+		// Backup the inventory state for restoration if needed.
+		m_pBackupUpgradeInv[i] = nullptr;
+		if (m_pMyUpgradeInv[i])
+		{
+			m_pBackupUpgradeInv[i] = new __IconItemSkill(*m_pMyUpgradeInv[i]);
+		}
+	}
+}
+
+void CUIItemUpgrade::Close()
+{
+	if (IsVisible())
+		SetVisible(false);
+	RestoreInventoryFromBackup();
+	AnimClose();
+
+	if (GetState() == UI_STATE_ICON_MOVING)
+		IconRestore();
+	SetState(UI_STATE_COMMON_NONE);
+	CN3UIWndBase::AllHighLightIconFree();
+
+	ItemMoveFromThisToInv();
+}
+
+void CUIItemUpgrade::ItemMoveFromThisToInv()
+{
+	CUIInventory* pInven = CGameProcedure::s_pProcMain->m_pUIInventory;
+	if (!pInven)
+		return;
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (m_pMyUpgradeInv[i])
+		{
+			__IconItemSkill* spItem = m_pMyUpgradeInv[i];
+			spItem->pUIIcon->SetParent(pInven);
+
+			m_pMyUpgradeInv[i] = nullptr;
+
+			CN3UIArea* pArea;
+
+			pArea = pInven->GetChildAreaByiOrder(UI_AREA_TYPE_INV, i);
+			if (pArea)
+			{
+				spItem->pUIIcon->SetRegion(pArea->GetRegion());
+				spItem->pUIIcon->SetMoveRect(pArea->GetRegion());
+			}
+			pInven->m_pMyInvWnd[i] = spItem;
+		}
+	}
+}
+
+bool CUIItemUpgrade::ReceiveIconDrop(__IconItemSkill* spItem, POINT ptCur)
+{
+	// Temp Define 
+#define FAIL_RETURN {	\
+		CN3UIWndBase::AllHighLightIconFree();	\
+		SetState(UI_STATE_COMMON_NONE);	\
+		return false;	\
+	}
+
+	CN3UIArea* pArea;
+	e_UIWND_DISTRICT eUIWnd = UIWND_DISTRICT_UNKNOWN;
+	if (!m_bVisible)
+		return false;
+
+	//  Check if the selected window is correct and the drop is valid.
+	if (CN3UIWndBase::m_sSelectedIconInfo.UIWndSelect.UIWnd != m_eUIWnd)
+		FAIL_RETURN
+		if ((CN3UIWndBase::m_sSelectedIconInfo.UIWndSelect.UIWndDistrict != UIWND_DISTRICT_UPGRADE_SLOT) &&
+			(CN3UIWndBase::m_sSelectedIconInfo.UIWndSelect.UIWndDistrict != UIWND_DISTRICT_UPGRADE_INV))
+			FAIL_RETURN
```

Comment:
Just `return false;` directly, and fix up the broken indentation caused by the macro.